File: zad2/zad2b.cpp
#include <iostream>
#include <vector>
#include <time.h>

using namespace std;

static long TimeCnt = 0;
static long ItCnt = 0;
static long ZBIR = 711;
static long PROIZVOD = 711000000;

/*
* ISPIS CENA PROIZVODA
*/
static void print(int x1, int x2, int x3, int x4) {
	cout
		<< "Prvi proizvod: "
		<< (double)x1 / 100
		<< endl
		<< "Drugi proizvod: "
		<< (double)x2 / 100
		<< endl
		<< "Treci proizvod: "
		<< (double)x3 / 100
		<< endl
		<< "Cetvrti proizvod: "
		<< (double)x4 / 100
		<< endl;
}

/*
* ISPIS BROJA ITERACIJA I PROTEKLOG VREMENA
*/
static void printStat() {
	cout
		<< "Broj iteracija: "
		<< ItCnt
		<< endl;

	cout
		<< "Vreme izvrsavanja: "
		<< ((double)clock() - TimeCnt) / CLOCKS_PER_SEC
		<< endl
		<< endl
		<< endl;
}


void func1(long zbir, long proizvod) {
	TimeCnt = clock();
	ItCnt = 0;

	for (long x1 = 1; x1 < zbir; x1++) {
		for (long x2 = x1; x2 < zbir; x2++) {
			for (long x3 = x2; x3 < zbir; x3++) {
				for (long x4 = x3; x4 < zbir; x4++) {
					ItCnt++;
					if ((x1 + x2 + x3 + x4 > zbir) || (x1 * x2 * x3 * x4) > proizvod)
						break;
					if ((x1 + x2 + x3 + x4 == zbir) && (x1 * x2 * x3 * x4 == proizvod))
						print(x1, x2, x3, x4);
				}
			}
		}
	}
	printStat();
}

void func2(long zbir, long proizvod) {
	TimeCnt = clock();
	ItCnt = 0;

	for (long x1 = 1; x1 < zbir; x1++) {
		for (long x2 = x1; x2 < zbir; x2++) {
			for (long x3 = x2; x3 < zbir; x3++) {
				ItCnt++;
				long x4 = zbir - x1 - x2 - x3;
				if ((x1 * x2 * x3 * x4) > proizvod)
					break;
				if ((x1 * x2 * x3 * x4) == proizvod)
					print(x1, x2, x3, x4);
			}
		}
	}
	printStat();
}

void func3(long zbir, long proizvod) {
	TimeCnt = clock();
	ItCnt = 0;
	vector<long> factors;
	for (int f = 2; f < zbir; f++) {
		if (proizvod % f == 0) {
			factors.push_back(f);
		}
	}
	for (int a = 0; a < (int)factors.size() - 3; a++) {
		for (int b = a + 1; b < (int)factors.size() - 2; b++) {
			for (int c = b + 1; c < (int)factors.size() - 1; c++) {
				for (int d = c + 1; d < (int)factors.size(); d++) {
					ItCnt++;
					if ((factors[a] + factors[b] + factors[c] + factors[d] > ZBIR) || (factors[a] * factors[b] * factors[c] * factors[d] > PROIZVOD))
						break;
					if ((factors[a] + factors[b] + factors[c] + factors[d] == ZBIR) && (factors[a] * factors[b] * factors[c] * factors[d] == PROIZVOD)) {
						print(factors[a], factors[b], factors[c], factors[d]);
					}
				}
			}
		}
	}
	printStat();
}

int main() {
	cout
		<< "--- Prvo resenje ---"
		<< endl
		<< endl;
	func1(ZBIR, PROIZVOD);

	cout
		<< "--- Drugo resenje ---"
		<< endl
		<< endl;
	func2(ZBIR, PROIZVOD);

	cout
		<< "--- Trece resenje ---"
		<< endl
		<< endl;
	func3(ZBIR, PROIZVOD);

	system("pause");
	return 0;
}

File: zad2/zad2b.py
import numpy as np
import math
from scipy.optimize import minimize

########## PRINT #####
def printFunc(x):
    x1, x2, x3, x4 = x
    
    print("Prvi proizvod: ", round(x1/100,2))
    print("Drugi proizvod: ", round(x2/100,2))
    print("Treci proizvod: ", round(x3/100,2))
    print("Cetvrti proizvod: ", round(x4/100,2))


########## FUN #####
def func1(x, y):
    x1, x2, x3, x4 = x
    cnt = 0
    S = y * 100
    P = y * 100000000
    for i1 in range(len(x1)):
        for i2 in range(len(x2)):
            for i3 in range(len(x3)):
                for i4 in range(len(x4)):
                    cnt = cnt + 1
                    a = x1[i1]
                    b = x2[i2]
                    c = x3[i3]
                    d = x4[i4]
                    if a+b+c+d > S:
                        break
                    if a*b*c*d > P:
                        break
                    if a+b+c+d == S and a+b+c+d == P:
                        print("Iterations: ", cnt)
                        return [a, b, c, d]
    print("Iterations: ", cnt)
    return [-1, -1, -1, -1]
        
def func2(x, y):
    x1, x2, x3 = x
    x4 = 0
    cnt = 0
    S = y * 100
    P = y * 100000000
    for i1 in range(len(x1)):
        for i2 in range(len(x2)):
            for i3 in range(len(x3)):
                cnt = cnt + 1
                a = x1[i1]
                b = x2[i2]
                c = x3[i3]
                x4 = S - a - b - c
                if a*b*c*x4 > P:
                    break
                if a*b*c*x4 == P:
                    print("Iterations: ", cnt)
                    return [a, b, c, x4]
    print("Iterations: ", cnt)
    return [-1, -1, -1, -1]
        
def func3(y):
    fac = []
    cnt = 0
    S = round(y * 100)
    P = round(y * 100000000)
    for i in range(2, S):
        if P % i == 0:
            fac.append(i)
            
    for i1 in range(len(fac) - 3):
        for i2 in range(len(fac) - 2):
            for i3 in range(len(fac) - 1):
                for i4 in range(len(fac)):
                    cnt = cnt + 1
                    a = fac[i1]
                    b = fac[i2]
                    c = fac[i3]
                    d = fac[i4]
                    if a+b+c+d > S or a*b*c*d > P:
                        break
                    if a+b+c+d == S and a*b*c*d == P:
                        print("Iterantions: ", cnt)
                        return [a, b, c, d]
    print("Iterations: ", cnt)
    return[-1, -1, -1, -1]
    
                        


########## MAIN #####
x = np.arange(1, 711, 1)

#####
print("3rd soultion:")
rs = func3(7.11)
printFunc(rs)
print("")
print("")
#####
print("2nd solution:")
rs = func2([x, x, x], 7.11)
printFunc(rs)
print("")
print("")
#####
print("1st solution:")
rs = func1([x, x, x, x], 7.11)
printFunc(rs)
print("")
print("")

